src/server/scripts/Outland/zone_shadowmoon_valley.cpp

```diff
@@ -571,64 +589,103 @@ class npc_dragonmaw_peon : public CreatureScript
+        {
+            events.ScheduleEvent(EVENT_KICK, urand(5000, 10000));
+            events.ScheduleEvent(EVENT_SUNDER, urand(5000, 10000));
         }
 
         void SpellHit(Unit* caster, const SpellInfo* spell)
         {
             if (!caster)
                 return;
 
-            if (caster->GetTypeId() == TYPEID_PLAYER && spell->Id == 40468 && !Tapped)
-            {
-                PlayerGUID = caster->GetGUID();
+            PlayerGUID = caster->GetGUID();
 
-                Tapped = true;
-                float x, y, z;
+            if (caster->GetTypeId() == TYPEID_PLAYER && spell->Id == SPELL_POISON)
+            {
                 caster->GetClosePoint(x, y, z, me->GetObjectSize());
-
-                me->SetWalk(false);
-                me->GetMotionMaster()->MovePoint(1, x, y, z);
+                Talk(SAY_1);
+                events2.ScheduleEvent(EVENT_WALK_TO_MUTTON, 0);
             }
         }
 
-        void MovementInform(uint32 type, uint32 id)
+        void MovementInform(uint32 /*type*/, uint32 id)
         {
-            if (type != POINT_MOTION_TYPE)
-                return;
+            if (id == 1)
+            {
+                me->HandleEmoteCommand(EMOTE_ONESHOT_EAT);
+                events2.ScheduleEvent(EVENT_POISIONED, 5000);
+            }
+        }
 
-            if (id)
+        void CreditPlayer()
+        {
+            if (PlayerGUID)
             {
-                me->SetUInt32Value(UNIT_NPC_EMOTESTATE, EMOTE_ONESHOT_EAT);
-                PoisonTimer = 15000;
+                Player* player = ObjectAccessor::GetPlayer(*me, PlayerGUID);
+                if (player && player->GetQuestStatus(11020) == QUEST_STATUS_INCOMPLETE)
+                    player->KilledMonsterCredit(23209, 0);
             }
         }
 
         void UpdateAI(uint32 diff)
         {
-            if (PoisonTimer)
+            events.Update(diff);
+            events2.Update(diff);
+
+            if (!UpdateVictim())
             {
-                if (PoisonTimer <= diff)
+                switch (events2.ExecuteEvent())
                 {
-                    if (PlayerGUID)
-                    {
-                        Player* player = ObjectAccessor::GetPlayer(*me, PlayerGUID);
-                        if (player && player->GetQuestStatus(11020) == QUEST_STATUS_INCOMPLETE)
-                            player->KilledMonsterCredit(23209, 0);
-                    }
-                    PoisonTimer = 0;
+                case EVENT_WALK_TO_MUTTON:
+                    me->SetWalk(true);
+                    me->GetMotionMaster()->MovePoint(1, x, y, z, true);
+                    me->SetUInt32Value(UNIT_NPC_EMOTESTATE, EMOTE_STATE_NONE);
+                    me->HandleEmoteCommand(EMOTE_ONESHOT_CHEER);
+                    break;
+                case EVENT_POISIONED:
+                    if (roll_chance_i(20))
+                        Talk(SAY_POISONED_1);
+                    me->CastSpell(me, SPELL_VOMIT);
```

Comment:
- me->CastSpell(me, SPELL_VOMIT);
+ CreditPlayer();
+ me->CastSpell(me, SPELL_VOMIT);
ps:SPELL_VOMIT affects the reading of PlayerGUID.  So complete the task

Reply:
This doesn't make any sence